utils/src/main/kotlin/Utils.kt

```diff
@@ -394,6 +394,11 @@ fun String.fileSystemDecode(): String =
         // URLDecoder does decode "." and "*".
         java.net.URLDecoder.decode(this, "UTF-8")
 
+/**
+ * Return the string encoded for safe use as a file name or "unknown", if the string is empty.
+ */
+fun String.encodeOrUnknown(): String = this.fileSystemEncode().takeUnless { it.isBlank() } ?: "unknown"
```

Comment:
Following the book this refactoring and the bug fix should be in separate commits. But I won't insist on it.

Reply:
And while at it, we should be able to drop the explicit return type and the `this.` from here.